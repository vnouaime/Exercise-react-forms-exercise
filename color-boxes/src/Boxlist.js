import React, {useState} from 'react'
import {v4 as uuid} from 'uuid'
import { Box } from './Box'
import { NewBoxForm } from './NewBoxForm'

/**
 * Returns NewBoxForm and any box components that have been stored in state
 * 
 * Props: none
 * 
 * State: array of boxes
 */
const Boxlist = () => {
    const INITIAL_STATE = []

    const [boxes, setBoxes] = useState(INITIAL_STATE)
    const addBox = (newBox) => {
        setBoxes(boxes => [...boxes, {...newBox, id: uuid()}])
    }
    const deleteBox = (deleteBox) => {
        setBoxes(boxes => boxes.filter(box => box.id !== deleteBox))
    }

    return (
        <>
        <h1>Add a Colored Box!</h1>
        <NewBoxForm addBox={addBox}/>
        {boxes.map(({ id, backgroundColor, width, height }) => (
            <Box 
                key={id} 
                id={id} 
                backgroundColor={backgroundColor} 
                width={width} 
                height={height} 
                deleteBox={deleteBox}
            />
        ))}

        </>
    )
}

export { Boxlist }