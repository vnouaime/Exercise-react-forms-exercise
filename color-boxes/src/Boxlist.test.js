import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { Boxlist } from './Boxlist'

function addBox(boxList, backgroundColor="blue", width="100", height="100") {
    const backgroundColorInput = boxList.getByLabelText("Background Color")
    const widthInput = boxList.getByLabelText("Width (pixels)")
    const heightInput = boxList.getByLabelText("Height (pixels)")
    const submitBtn = boxList.queryByText("Add Box")
    fireEvent.change(backgroundColorInput, { target: { value: backgroundColor}})
    fireEvent.change(widthInput, { target: { value: width}})
    fireEvent.change(heightInput, { target: { value: height}})
    fireEvent.click(submitBtn)
}

test('smoke test of Boxlist', () => {
    render(<Boxlist />)
})

it("matches snapshot", function() {
    const {asFragment} = render(<Boxlist />);
    expect(asFragment()).toMatchSnapshot();
});

it("adds a new box", function() {
    const boxList = render(<Boxlist />)
    expect(boxList.queryByText("X")).not.toBeInTheDocument();
    addBox(boxList)

    const deleteButton = boxList.getByText("X");
    expect(deleteButton).toBeInTheDocument();
    expect(deleteButton.previousSibling).toHaveStyle(`
    background-color: blue;
    width: 100px;
    height: 100px;
  `);

    expect(boxList.getAllByDisplayValue("")).toHaveLength(3);
})

it("removes a box", function() {
    const boxList = render(<Boxlist />);
    addBox(boxList);

    const deleteButton = boxList.getByText("X");
    fireEvent.click(deleteButton);
    expect(deleteButton).not.toBeInTheDocument();
})
