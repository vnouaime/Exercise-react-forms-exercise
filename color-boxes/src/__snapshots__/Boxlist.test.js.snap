// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <h1>
    Add a Colored Box!
  </h1>
  <form>
    <label
      for="backgroundColor"
    >
      Background Color
    </label>
    <input
      id="backgroundColor"
      name="backgroundColor"
      placeholder="Background Color"
      type="text"
      value=""
    />
    <label
      for="width"
    >
      Width (pixels)
    </label>
    <input
      id="width"
      name="width"
      placeholder="Width"
      type="text"
      value=""
    />
    <label
      for="height"
    >
      Height (pixels)
    </label>
    <input
      id="height"
      name="height"
      placeholder="Height"
      type="text"
      value=""
    />
    <button
      class="NewBoxForm-submitButton"
    >
      Add Box
    </button>
  </form>
</DocumentFragment>
`;
